Add placeholder mitigation content to rest api docs (#277)
alphagov/notifications-tech-docs: source/documentation/_api_docs.md

```
You can leave out this argument if a template does not have any placeholder fields for personalised information.

> Find out how to reduce the risk of malicious content injection in placeholders [here](#reducing-the-risk-of-malicious-content-injection-in-placeholders) 

Reviewer: Don’t use ‘here’ as link text. Meaningful link text is [important for accessibility](https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context).
```suggestion
Find out how to [reduce the risk of malicious content injection in placeholders](#reducing-the-risk-of-malicious-content-injection-in-placeholders).
```

Author: Thanks for the link to the link purpose success criterion.

}
```

#### Reducing the risk of malicious content injection in placeholders

Notify lets you [personalise messages](https://www.notifications.service.gov.uk/using-notify/personalisation) using placeholders.

You can [format](https://www.notifications.service.gov.uk/using-notify/formatting) content or add links and urls into placeholders using Markdown. 

Information from untrusted sources, such as online forms may be used to add malicious content and links. You might pass this information into Notify personalisation. 

The malicious content could be:

  * Markdown syntax intended to be rendered into HTML 
  * a plain text URL which would be rendered into a clickable phishing link
  * an example of malicious content used in Notify personalisation:

An example of how malicious content can be injected into Notify personalisation:

> *Template in Notify:*

Reviewer 2: ```suggestion
Template in Notify:
```

Could we remove the italics as it makes it hard to read / understand. There's examples of a format to use in [REST API](https://docs.notifications.service.gov.uk/rest-api.html#authorisation-header)

Author: Yes, that is a good point. I will update it

  ```
  Hello ((name))
  ```

> *Personalisation:* 

Reviewer 2: Remove italics

  ```javascript
    {name: "Anne Example, now [click this evil link](https://malicious.link)"}
  ```

> *email will appear as:* 

Reviewer 2: remove italics

 ```
    Hello Anne Example, now [click this evil link](https://malicious.link)
 ```

We recommend you sanitise all input from untrusted sources to prevent the injection of malicious content. You can:

> 1. Use the Markdown triple backtick [fenced code blocks](https://www.markdownguide.org/extended-syntax/#fenced-code-blocks) syntax

Reviewer: Single or triple backticks are escapable at the moment. If my malicious content looks like this then I can escape:
```markdown
    ```
    [evil link](https://example.com) 
    ```
```

Author: I have added more context i.e that it is a solution that is vulnerable to being nullified by the addition of triple backticks.
I have also changed the order so that the solution of escaping with a backslash is the main tip being given.

It may be that we drop it as a tip altogether.

      ```
          ((name))
      ```

>     The triple backtick won’t be displayed in the email notification and any markdown content or URL link injected into the placeholder will not be rendered.

Reviewer 2: Could we add the syntax in a grey box like in [this example](https://docs.notifications.service.gov.uk/rest-api.html#base-url)?

> 1. Use a backslash to escape [individual characters](https://www.markdownguide.org/basic-syntax/#characters-you-can-escape) such as “\”, “<”, “+”.

Reviewer: `<` is not a relevant character

The most dangerous ones are those for links (`[]()`)

}
```

#### Reducing the risk of malicious content injection in placeholders

Notify lets you [personalise messages](https://www.notifications.service.gov.uk/using-notify/personalisation) using placeholders.

You can [format](https://www.notifications.service.gov.uk/using-notify/formatting) content or add links and urls into placeholders using Markdown. 

Information from untrusted sources, such as online forms may be used to add malicious content and links. You might pass this information into Notify personalisation. 

The malicious content could be:

  * Markdown syntax intended to be rendered into HTML 
  * a plain text URL which would be rendered into a clickable phishing link

>   * an example of malicious content used in Notify personalisation:

Author: Yes that was a mistake.

An example of how malicious content can be injected into Notify personalisation:

**Template in Notify**:
  ```
  Hello ((name))
  ```

**Personalisation**:
  ```javascript
    {name: "Anne Example, now [click this evil link](https://malicious.link)"}
  ```

> **email will appear as**:

Reviewer 2: ```suggestion
**Email will appear as**:
```

 ```

>     Hello Anne Example, now [click this evil link](https://malicious.link)

Author: I will look into it, anything that helps to demonstrate the issue is most welcome.

 ```

We recommend you sanitise all input from untrusted sources to prevent the injection of malicious content. You can:

1. Use a backslash to escape [individual characters](https://www.markdownguide.org/basic-syntax/#characters-you-can-escape). 
   The characters of most concern are those that could be used to add a URL link such as: “\[”, “\]”, “\(”, "\)".

2. Use the Markdown triple backtick [fenced code blocks](https://www.markdownguide.org/extended-syntax/#fenced-code-blocks) syntax

>    to escape multiple characters. This is a less effective option as it can be overcome by a would be attacker preceding malicious content with triple backticks.

Reviewer 3: ```suggestion
   to escape all characters within a placeholder. Keep in mind that to use this option, the placeholder needs to be on a separate line in the template.

This is also a less effective option - an attacker could potentially overcome it by preceding malicious content with triple backticks of their own.
```

You can [format](https://www.notifications.service.gov.uk/using-notify/formatting) content or add links and urls into placeholders using Markdown. 

If you pass in information from untrusted sources (such as online forms) into your Notify template using personalisation, this may be used to add malicious content and links to notifications you send via Notify. 

The malicious content could be:

  * Markdown syntax intended to be rendered into HTML 
  * a plain text URL which would be rendered into a clickable phishing link

An example of how malicious content can be injected into Notify personalisation:

**Template in Notify**:

```
Hello ((name))
```

**Personalisation**: 

```
{name: "Anne Example, now [click this evil link](https://malicious.link)"}
```

**Email will appear as**: 

<pre>
 <small>Dear Anne Example, now <a href="https://malicious.link>click this evil link">click this evil link</a></small>
</pre>

We recommend you sanitise all input from untrusted sources to prevent the injection of malicious content. You can:

1. Use a backslash to escape [individual characters](https://www.markdownguide.org/basic-syntax/#characters-you-can-escape). 

>    The characters of most concern are those that could be used to add a URL link such as: “\[”, “\]”, “\(”, "\)".

Reviewer 3: I think this could look clearer? The paranetheses are also characters, so including them could be visually confusing.

```suggestion
   The characters of most concern are those that could be used to add a URL link such as: `[`, `]`, `(` or `)`.
```

2. Use the Markdown triple backtick [fenced code blocks](https://www.markdownguide.org/extended-syntax/#fenced-code-blocks) syntax
   to escape all characters within a placeholder. 
   Keep in mind that to use this option, the placeholder needs to be on a separate line from the backticks in the template. 
   It is also a less effective option - an attacker could potentially overcome it by preceding malicious content with triple backticks of their own.
   For it to be effective, the input must be pre-processed to escape triple backticks.

~~~
```
((name))
```
~~~

> 

Reviewer 3: I actually also lean towards not having the backticks option in the guidance - it's not very practical.

> 

Author: It is easier to take it out, so I will take it out.
I will wait till tomorrow to see if there are any other suggestions and will then open a PR to update the other docs.
